In [25]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [26]:
df = pd.read_csv('./data/Life-Expectancy-Data-Averaged.csv')
df

Country                         Region    Year  \
0            Afghanistan                           Asia  2007.5   
1                Albania                 Rest of Europe  2007.5   
2                Algeria                         Africa  2007.5   
3                 Angola                         Africa  2007.5   
4    Antigua and Barbuda  Central America and Caribbean  2007.5   
..                   ...                            ...     ...   
174        Venezuela, RB                  South America  2007.5   
175              Vietnam                           Asia  2007.5   
176          Yemen, Rep.                    Middle East  2007.5   
177               Zambia                         Africa  2007.5   
178             Zimbabwe                         Africa  2007.5   

     Infant_deaths  Under_five_deaths  Adult_mortality  Alcohol_consumption  \
0         71.08125           98.61250       265.804969             0.016125   
1         15.25625           17.14375        83.132969             4.696875   
2         26.75625           31.19375       113.439281             0.400625   
3         88.76875          144.16250       297.844063             4.935625   
4          9.47500           11.51875       142.478813             7.755000   
..             ...                ...              ...                  ...   
174       15.85000           18.50000       151.056156             7.458125   
175       19.50000           24.47500       133.054406             1.742500   
176       52.44375           69.57500       239.653938             0.047000   
177       60.58125           99.20625       452.761219             2.903125   
178       50.16875           85.32500       572.974312             2.900625   

     Hepatitis_B  Measles       BMI    Polio  Diphtheria  Incidents_HIV  \
0        64.5625  24.3750  22.46250  55.3750     55.1250       0.022500   
1        98.0000  95.9375  25.85625  98.1250     98.0625       0.025625   
2        88.3125  93.2500  24.86875  91.7500     91.8750       0.021875   
3        68.8125  64.0000  22.51875  35.7500     55.5625       1.303750   
4        98.2500  75.4375  25.85000  96.9375     98.3125       0.125000   
..           ...      ...       ...      ...         ...            ...   
174      72.4375  83.0000  26.49375  79.1875     75.2500       0.400000   
175      87.4375  65.0000  20.76250  94.9375     91.7500       0.196875   
176      60.1875  95.0000  23.02500  71.6250     72.6250       0.025000   
177      82.8750  39.1875  22.00000  82.9375     83.8125       6.868125   
178      79.5625  64.0000  23.54375  79.5625     79.1250       8.006250   

     GDP_per_capita  Population_mln  Thinness_ten_nineteen_years  \
0          408.5625       27.450625                     16.58125   
1         3071.1250        2.969375                      1.61875   
2         3745.1250       34.820625                      6.09375   
3         2647.8125       21.623750                      6.19375   
4        14678.7500        0.085000                      3.42500   
..              ...             ...                          ...   
174       9294.8750       27.392500                      1.65000   
175       1807.8750       86.032500                     14.92500   
176       2406.8125       21.730625                     13.83125   
177       1059.8750       12.862500                      6.88125   
178       1290.6250       12.538750                      7.01250   

     Thinness_five_nine_years  Schooling  Economy_status  Life_expectancy  
0                    15.58125    2.90000             1.0         59.65625  
1                     1.70000    9.24375             1.0         75.95000  
2                     5.97500    6.99375             1.0         73.78750  
3                     6.66875    4.60625             1.0         52.82500  
4                     3.37500    9.01875             1.0         75.35000  
..                        ...        ...             ...              ...  
174          

In [27]:
df.isnull().sum()

Country                        0
Region                         0
Year                           0
Infant_deaths                  0
Under_five_deaths              0
Adult_mortality                0
Alcohol_consumption            0
Hepatitis_B                    0
Measles                        0
BMI                            0
Polio                          0
Diphtheria                     0
Incidents_HIV                  0
GDP_per_capita                 0
Population_mln                 0
Thinness_ten_nineteen_years    0
Thinness_five_nine_years       0
Schooling                      0
Economy_status                 0
Life_expectancy                0
dtype: int64

- 결측값이 없음
- 각 나라별 데이터므로 이상치 처리는 따로 하지 않고 진행

In [28]:
df.describe()

Year  Infant_deaths  Under_five_deaths  Adult_mortality  \
count   179.0     179.000000         179.000000       179.000000   
mean   2007.5      30.363792          42.938268       192.251775   
std       0.0      26.725485          42.916952       111.659044   
min    2007.5       2.381250           3.000000        57.710313   
25%    2007.5       8.159375           9.775000       107.046906   
50%    2007.5      19.368750          23.137500       164.432406   
75%    2007.5      48.959375          68.321875       247.523922   
max    2007.5     115.718750         178.725000       572.974312   

       Alcohol_consumption  Hepatitis_B     Measles         BMI       Polio  \
count           179.000000   179.000000  179.000000  179.000000  179.000000   
mean              4.820882    84.292598   77.344972   25.032926   86.499651   
std               3.914554    13.820223   17.315208    2.165490   13.581153   
min               0.000025    30.687500   16.250000   20.212500   35.750000   
25%               1.317813    78.218750   64.000000   23.225000   80.531250   
50%               4.209375    88.000000   83.000000   25.650000   92.375000   
75%               7.843438    94.375000   92.250000   26.425000   96.062500   
max              15.100000    98.875000   99.000000   31.687500   98.937500   

       Diphtheria  Incidents_HIV  GDP_per_capita  Population_mln  \
count  179.000000     179.000000      179.000000      179.000000   
mean    86.271648       0.894288    11540.924930       36.675915   
std     13.931532       2.311895    16893.054182      136.655286   
min     31.312500       0.010000      263.937500        0.085000   
25%     80.812500       0.080000     1409.906250        2.108125   
50%     92.062500       0.164375     4402.625000        7.660625   
75%     95.781250       0.516250    12037.781250       22.745313   
max     99.000000      18.164375   102972.687500     1321.239375   

       Thinness_ten_nineteen_years  Thinness_five_nine_years   Schooling  \
count                   179.000000                179.000000  179.000000   
mean                      4.865852                  4.899825    7.632123   
std                       4.111094                  4.195663    3.126912   
min                       0.100000                  0.100000    1.337500   
25%                       1.756250                  1.731250    4.946875   
50%                       3.556250                  3.718750    7.831250   
75%                       7.165625                  7.056250   10.365625   
max                      27.100000                 27.943750   13.268750   

       Economy_status  Life_expectancy  
count      179.000000       179.000000  
mean         0.793296        68.856075  
std          0.406077         9.197699  
min          0.000000        45.606250  
25%          1.000000        62.303125  
50%          1.000000        71.506250  
75%          1.000000        74.937500  
max          1.000000        82.456250

- 값의 범위가 큰 컬럼에 대해 스케일링 진행

In [29]:
from sklearn.preprocessing import MinMaxScaler

# MinMaxScaler 객체 생성
scaler = MinMaxScaler()

scale_column = ['Infant_deaths', 'Under_five_deaths', 'Adult_mortality', 'GDP_per_capita', 'Population_mln']


df[scale_column] = scaler.fit_transform(df[scale_column])
df


Country                         Region    Year  \
0            Afghanistan                           Asia  2007.5   
1                Albania                 Rest of Europe  2007.5   
2                Algeria                         Africa  2007.5   
3                 Angola                         Africa  2007.5   
4    Antigua and Barbuda  Central America and Caribbean  2007.5   
..                   ...                            ...     ...   
174        Venezuela, RB                  South America  2007.5   
175              Vietnam                           Asia  2007.5   
176          Yemen, Rep.                    Middle East  2007.5   
177               Zambia                         Africa  2007.5   
178             Zimbabwe                         Africa  2007.5   

     Infant_deaths  Under_five_deaths  Adult_mortality  Alcohol_consumption  \
0         0.606154           0.544103         0.403860             0.016125   
1         0.113599           0.080488         0.049339             4.696875   
2         0.215066           0.160442         0.108156             0.400625   
3         0.762215           0.803315         0.466040             4.935625   
4         0.062590           0.048478         0.164515             7.755000   
..             ...                ...              ...                  ...   
174       0.118838           0.088206         0.181161             7.458125   
175       0.151042           0.122208         0.146224             1.742500   
176       0.441712           0.378859         0.353108             0.047000   
177       0.513511           0.547482         0.766696             2.903125   
178       0.421639           0.468488         1.000000             2.900625   

     Hepatitis_B  Measles       BMI    Polio  Diphtheria  Incidents_HIV  \
0        64.5625  24.3750  22.46250  55.3750     55.1250       0.022500   
1        98.0000  95.9375  25.85625  98.1250     98.0625       0.025625   
2        88.3125  93.2500  24.86875  91.7500     91.8750       0.021875   
3        68.8125  64.0000  22.51875  35.7500     55.5625       1.303750   
4        98.2500  75.4375  25.85000  96.9375     98.3125       0.125000   
..           ...      ...       ...      ...         ...            ...   
174      72.4375  83.0000  26.49375  79.1875     75.2500       0.400000   
175      87.4375  65.0000  20.76250  94.9375     91.7500       0.196875   
176      60.1875  95.0000  23.02500  71.6250     72.6250       0.025000   
177      82.8750  39.1875  22.00000  82.9375     83.8125       6.868125   
178      79.5625  64.0000  23.54375  79.5625     79.1250       8.006250   

     GDP_per_capita  Population_mln  Thinness_ten_nineteen_years  \
0          0.001408        0.020713                     16.58125   
1          0.027332        0.002183                      1.61875   
2          0.033894        0.026292                      6.09375   
3          0.023210        0.016303                      6.19375   
4          0.140346        0.000000                      3.42500   
..              ...             ...                          ...   
174        0.087928        0.020669                      1.65000   
175        0.015032        0.065055                     14.92500   
176        0.020864        0.016384                     13.83125   
177        0.007749        0.009671                      6.88125   
178        0.009996        0.009426                      7.01250   

     Thinness_five_nine_years  Schooling  Economy_status  Life_expectancy  
0                    15.58125    2.90000             1.0         59.65625  
1                     1.70000    9.24375             1.0         75.95000  
2                     5.97500    6.99375             1.0         73.78750  
3                     6.66875    4.60625             1.0         52.82500  
4                     3.37500    9.01875             1.0         75.35000  
..                        ...        ...             ...              ...  
174          

In [30]:
print(df['Country'].unique())
print(df['Region'].unique())

['Afghanistan' 'Albania' 'Algeria' 'Angola' 'Antigua and Barbuda'
 'Argentina' 'Armenia' 'Australia' 'Austria' 'Azerbaijan' 'Bahamas, The'
 'Bahrain' 'Bangladesh' 'Barbados' 'Belarus' 'Belgium' 'Belize' 'Benin'
 'Bhutan' 'Bolivia' 'Bosnia and Herzegovina' 'Botswana' 'Brazil'
 'Brunei Darussalam' 'Bulgaria' 'Burkina Faso' 'Burundi' 'Cabo Verde'
 'Cambodia' 'Cameroon' 'Canada' 'Central African Republic' 'Chad' 'Chile'
 'China' 'Colombia' 'Comoros' 'Congo, Dem. Rep.' 'Congo, Rep.'
 'Costa Rica' "Cote d'Ivoire" 'Croatia' 'Cuba' 'Cyprus' 'Czechia'
 'Denmark' 'Djibouti' 'Dominican Republic' 'Ecuador' 'Egypt, Arab Rep.'
 'El Salvador' 'Equatorial Guinea' 'Eritrea' 'Estonia' 'Eswatini'
 'Ethiopia' 'Fiji' 'Finland' 'France' 'Gabon' 'Gambia, The' 'Georgia'
 'Germany' 'Ghana' 'Greece' 'Grenada' 'Guatemala' 'Guinea' 'Guinea-Bissau'
 'Guyana' 'Haiti' 'Honduras' 'Hungary' 'Iceland' 'India' 'Indonesia'
 'Iran, Islamic Rep.' 'Iraq' 'Ireland' 'Israel' 'Italy' 'Jamaica' 'Japan'
 'Jordan' 'Kazakhstan' 'K

In [31]:
# 나라, 지역 종류가 너무 많아 제거 / 년도는 전부 동일해서 같이 제거
df = df.drop(columns=["Country", "Region", "Year"])
df

Infant_deaths  Under_five_deaths  Adult_mortality  Alcohol_consumption  \
0         0.606154           0.544103         0.403860             0.016125   
1         0.113599           0.080488         0.049339             4.696875   
2         0.215066           0.160442         0.108156             0.400625   
3         0.762215           0.803315         0.466040             4.935625   
4         0.062590           0.048478         0.164515             7.755000   
..             ...                ...              ...                  ...   
174       0.118838           0.088206         0.181161             7.458125   
175       0.151042           0.122208         0.146224             1.742500   
176       0.441712           0.378859         0.353108             0.047000   
177       0.513511           0.547482         0.766696             2.903125   
178       0.421639           0.468488         1.000000             2.900625   

     Hepatitis_B  Measles       BMI    Polio  Diphtheria  Incidents_HIV  \
0        64.5625  24.3750  22.46250  55.3750     55.1250       0.022500   
1        98.0000  95.9375  25.85625  98.1250     98.0625       0.025625   
2        88.3125  93.2500  24.86875  91.7500     91.8750       0.021875   
3        68.8125  64.0000  22.51875  35.7500     55.5625       1.303750   
4        98.2500  75.4375  25.85000  96.9375     98.3125       0.125000   
..           ...      ...       ...      ...         ...            ...   
174      72.4375  83.0000  26.49375  79.1875     75.2500       0.400000   
175      87.4375  65.0000  20.76250  94.9375     91.7500       0.196875   
176      60.1875  95.0000  23.02500  71.6250     72.6250       0.025000   
177      82.8750  39.1875  22.00000  82.9375     83.8125       6.868125   
178      79.5625  64.0000  23.54375  79.5625     79.1250       8.006250   

     GDP_per_capita  Population_mln  Thinness_ten_nineteen_years  \
0          0.001408        0.020713                     16.58125   
1          0.027332        0.002183                      1.61875   
2          0.033894        0.026292                      6.09375   
3          0.023210        0.016303                      6.19375   
4          0.140346        0.000000                      3.42500   
..              ...             ...                          ...   
174        0.087928        0.020669                      1.65000   
175        0.015032        0.065055                     14.92500   
176        0.020864        0.016384                     13.83125   
177        0.007749        0.009671                      6.88125   
178        0.009996        0.009426                      7.01250   

     Thinness_five_nine_years  Schooling  Economy_status  Life_expectancy  
0                    15.58125    2.90000             1.0         59.65625  
1                     1.70000    9.24375             1.0         75.95000  
2                     5.97500    6.99375             1.0         73.78750  
3                     6.66875    4.60625             1.0         52.82500  
4                     3.37500    9.01875             1.0         75.35000  
..                        ...        ...             ...              ...  
174                   1.56250    8.23750             1.0         72.79375  
175                  15.62500    6.90000             1.0         74.33125  
176                  13.75000    2.22500             1.0         64.16250  
177                   6.76250    6.43750             1.0         52.33125  
178                   6.98750    7.21250             1.0         48.66250  

[179 rows x 17 columns]

In [32]:
# 훈련-테스트 데이터 분할 후 스케일링
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

y = df["Life_expectancy"]
X = df.drop(columns=["Life_expectancy"])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()


X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


### 1. 선형 모델링

In [33]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# 모델 생성
lr = LinearRegression()

# 모델 학습
lr.fit(X_train, y_train)

# 예측
y_pred_test = lr.predict(X_test)

# 성능 평가
mse = mean_squared_error(y_test, y_pred_test)
mae = mean_absolute_error(y_test, y_pred_test)
r2 = r2_score(y_test, y_pred_test)

print(f"Linear Regression MSE: {mse}")
print(f"Linear Regression MAE: {mae}")
print(f"Linear Regression R²: {r2}")




Linear Regression MSE: 1.1144170028881946
Linear Regression MAE: 0.8211173801618181
Linear Regression R²: 0.9875885250502776


### 2. 랜덤 포레스트

In [34]:
from sklearn.ensemble import RandomForestRegressor

# 모델 생성
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)

# 모델 학습
rf_model.fit(X_train, y_train)

# 예측
y_pred_rf = rf_model.predict(X_test)

# 성능 평가
mse_rf = mean_squared_error(y_test, y_pred_rf)
mae_rf = mean_absolute_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print(f"Random Forest MSE: {mse_rf}")
print(f"Random Forest MAE: {mae_rf}")
print(f"Random Forest R²: {r2_rf}")


Random Forest MSE: 2.143880905924489
Random Forest MAE: 1.1665052083333427
Random Forest R²: 0.9761231889946861


### 3. XGBoost

In [35]:
from xgboost import XGBRegressor

# 모델 생성
xgb_model = XGBRegressor(n_estimators=100, random_state=42)

# 모델 학습
xgb_model.fit(X_train, y_train)

# 예측
y_pred_xgb = xgb_model.predict(X_test)

# 성능 평가
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
mae_xgb = mean_absolute_error(y_test, y_pred_xgb)
r2_xgb = r2_score(y_test, y_pred_xgb)

print(f"XGBoost MSE: {mse_xgb}")
print(f"XGBoost MAE: {mae_xgb}")
print(f"XGBoost R²: {r2_xgb}")


XGBoost MSE: 2.807533702180808
XGBoost MAE: 1.3098873350355358
XGBoost R²: 0.9687319610838581


### 4. LightGBM

In [36]:
import lightgbm as lgb

# 모델 생성
lgb_model = lgb.LGBMRegressor(n_estimators=100, random_state=42)

# 모델 학습
lgb_model.fit(X_train, y_train)

# 예측
y_pred_lgb = lgb_model.predict(X_test)

# 성능 평가
mse_lgb = mean_squared_error(y_test, y_pred_lgb)
mae_lgb = mean_absolute_error(y_test, y_pred_lgb)
r2_lgb = r2_score(y_test, y_pred_lgb)

print(f"LightGBM MSE: {mse_lgb}")
print(f"LightGBM MAE: {mae_lgb}")
print(f"LightGBM R²: {r2_lgb}")


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000055 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 702
[LightGBM] [Info] Number of data points in the train set: 143, number of used features: 16
[LightGBM] [Info] Start training from score 68.849475
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, b

### 5. Neural Network

In [37]:
from sklearn.neural_network import MLPRegressor

# 모델 생성
nn_model = MLPRegressor(hidden_layer_sizes=(100,), max_iter=1000, random_state=42)

# 모델 학습
nn_model.fit(X_train, y_train)

# 예측
y_pred_nn = nn_model.predict(X_test)

# 성능 평가
mse_nn = mean_squared_error(y_test, y_pred_nn)
mae_nn = mean_absolute_error(y_test, y_pred_nn)
r2_nn = r2_score(y_test, y_pred_nn)

print(f"Neural Network MSE: {mse_nn}")
print(f"Neural Network MAE: {mae_nn}")
print(f"Neural Network R²: {r2_nn}")


Neural Network MSE: 148.5097121393771
Neural Network MAE: 10.404850177065123
Neural Network R²: -0.653981020777084


- Linear Regression MSE: 1.1144170028882103
- Linear Regression MAE: 0.821117380161826
- Linear Regression R²: 0.9875885250502775

- Random Forest MSE: 2.143880905924489
- Random Forest MAE: 1.1665052083333427
- Random Forest R²: 0.9761231889946861

- XGBoost MSE: 2.807533702180808
- XGBoost MAE: 1.3098873350355358
- XGBoost R²: 0.9687319610838581

- LightGBM MSE: 2.860741850371836
- LightGBM MAE: 1.3846531169341416
- LightGBM R²: 0.9681393717778061

- Neural Network MSE: 148.5097121393771
- Neural Network MAE: 10.404850177065123
- Neural Network R²: -0.653981020777084